File: Javascript/121_Best_Time_to_Buy_and_Sell_Stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    // let buy = prices[0]
    // let profit = 0
    // for(let i =0; i<prices.length;i++){
    //     if(prices[i]<buy){
    //         buy = prices[i]
    //     }
    //     if(prices[i]-buy>profit){
    //         profit = prices[i]-buy
    //     }
    // }
    // return profit
    let b1 = -Infinity
    let s1 = 0
    for(let i =0; i<prices.length;i++){
        b1= Math.max(b1,-prices[i])
        s1 = Math.max(s1,b1+prices[i])
    }
    return s1
};

File: python/121_Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min = prices[0]
        max = prices[0]
        profit = 0
        for price in prices:
            if price < min and price != prices[-1]:
                min = price
                max = price
            if  price > max:
                max = price
            if max-min > profit:
                profit = max-min
        return profit
